Программа А из четвёртого модуля осени 2019 года

DmitryInd opened the pull request.

DmitryInd committed:
Программа А из четвёртого модуля, посылка с id 29094722

DmitryInd committed:
Подправил комментарии

Ya-Egor commented on Contest 29_11_19/Program A/biginteger.h:
```diff
@@ -0,0 +1,519 @@
+}
+
+//Сравнение >= по модулю
+bool compare_abs(BigInteger left, BigInteger right) {
```
Кажется жутко неэффективно ради сравнения заводить копии операндов. Неужели обязательно их копировать?

DmitryInd replied:
Исправил

Ya-Egor commented on Contest 29_11_19/Program A/biginteger.h:
```diff
@@ -0,0 +1,519 @@
+		right_value += base * right.value[1];
+	}
+	BigInteger div;
+	div.value.push_back(left_value / right_value);
```
А при делении всегда получается одноразрядное число?

DmitryInd replied:
Да, так как с помощью этой функции считается только нижняя оценка деления (в операторе '/'), которая равна A/(B + 1).
left = A, right = B + 1
A - это старший разряд левого числа, B - старший правого
left = A * base + следующий разряд только, когда A < B + 1, следовательно, A' < B * base  + base, следовательно деление умещается в один разряд, ч. т. д.

DmitryInd committed:
Исправил функции сравнения
1. Убрано копирование для сравнения по модулю.
2. Теперь сравнение основано на сравнении по модулю, а не наоборот.
3. Исправлена грубая ошибка в обычном сравнении.
